internal/ent/hooks/event.go

```diff
@@ -0,0 +1,249 @@
+}
+
+// TODO: template out listeners for all mutation types and create hooks that allow functions to be easily called from within the codebase to perform actions on those events
+// RegisterListeners registers listeners for events on the event pool and is registered with the dbclient
+func RegisterListeners(pool *soiree.EventPool) {
+	_, err := pool.On("OrganizationSetting.OpCreate", handleCustomerCreate)
+	if err != nil {
+		log.Error().Err(ErrFailedToRegisterListener)
+	}
+
+	for _, event := range []string{"OrganizationSetting.OpUpdate", "OrganizationSetting.OpUpdateOne"} {
+		_, err = pool.On(event, handleCustomerCreate)
+		if err != nil {
+			log.Error().Err(ErrFailedToRegisterListener)
+		}
+	}
+}
+
+// handleCustomerCreate handles the creation of a customer in Stripe
+func handleCustomerCreate(event soiree.Event) error {
+	props := event.Properties()
+	// TODO: add funcs for default unmarshalling of props and send all props to stripe
+	orgsettingID, exists := props["ID"]
+	if !exists {
+		log.Info().Msg("organizationSetting ID not found in event properties")
+		return nil
+	}
+
+	// if stripe ID is being added to the properties, we enumerate through any other fields in the update and ensure they match
+	stripeID, exists := props["stripe_id"]
+	if exists && stripeID != "" {
+		stripes := stripeID.(string)
+
+		stripecustomer, err := event.Client().(*entgen.Client).EntitlementManager.Client.Customers.Get(stripes, nil)
+		if err != nil {
+			log.Err(err).Msg("Failed to retrieve Stripe customer by ID")
+			return err
+		}
+
+		billingEmail, exists := props["billing_email"]
+		if exists && billingEmail != "" {
+			email := billingEmail.(string)
+			if stripecustomer.Email != email {
+				_, err := event.Client().(*entgen.Client).EntitlementManager.Client.Customers.Update(stripes, &stripe.CustomerParams{
+					Email: &email,
+				})
+				if err != nil {
+					log.Err(err).Msg("failed to update Stripe customer email")
+					return err
+				}
+			}
+		}
+
+		billingPhone, exists := props["billing_phone"]
+		if exists && billingPhone != "" {
+			phone := billingPhone.(string)
+			if stripecustomer.Phone != phone {
+				_, err := event.Client().(*entgen.Client).EntitlementManager.Client.Customers.Update(stripes, &stripe.CustomerParams{
+					Phone: &phone,
+				})
+				if err != nil {
+					log.Err(err).Msg("failed to update Stripe customer phone")
+					return err
+				}
+			}
+		}
+
+		// TODO: split out address fields in ent schema
+		billingAddress, exists := props["billing_address"]
+		if exists && billingAddress != "" {
+			address := billingAddress.(string)
+			if stripecustomer.Address.Line1 != address {
+				_, err := event.Client().(*entgen.Client).EntitlementManager.Client.Customers.Update(stripes, &stripe.CustomerParams{
+					Address: &stripe.AddressParams{
+						Line1: &address,
+					},
+				})
+				if err != nil {
+					log.Err(err).Msg("failed to update Stripe customer address")
+					return err
+				}
+			}
+		}
```

Comment:
I think you need a return here, otherwise I think it will attempt to create another record if the `billing_email` is in the props?

Reply:
this is a psuedo reconciliation because if billing_email is present in the mutation, but the stripe ID does NOT match, we want to update the existing stripe record by ID with the email inside of the mutation. Because this is triggered when we first create the email in the settings and then have to handle updates to the email on an ongoing basis (and based on whether or not there is a stripe ID with an email that matches) we have to evaluate it both times